src/test/ui/intrinsics/is_const_eval_unleash.rs

```diff
@@ -0,0 +1,34 @@
+// run-pass
+// only-x86_64
+// compile-flags: -Zunleash-the-miri-inside-of-you
+
+#![feature(core_intrinsics)]
+use std::intrinsics::is_const_eval;
+use std::arch::x86_64::*;
+use std::mem::transmute;
+
+const fn eq(x: [i32; 4], y: [i32; 4]) -> bool {
+    if unsafe { is_const_eval() } {
+        x[0] == y[0] && x[1] == y[1] && x[2] == y[2] && x[3] == y[3]
+    } else {
+        unsafe {
+            let x = _mm_loadu_si128(&x as *const _ as *const _);
```

Comment:
This is essentially what I worried about. Such code will never ever be accepted by rustc AFAIK. If you want it for experimentation, that seems fine but ultimately pointless. Writing a const qualification that could accept this would be the motherload of unsoundness dangers. My suggestion with an intrinsic, that calls one of two functions given to it depending on the context, would be much more manageable

Reply:
Long version:

In contrast to C++, we statically prove that your const fn is sane. Of course we also have escape hatches via unsafe, but that's a different story. What your example has is an if condition that works differently whether you are const evaluating or codegenning. Now, this is not a problem by itself, but the problem arises with the `else` arm's content. That content contains statically provable nonconst code and since it's part of a const fn that will be errored about. I'm presuming the intent is to decide whether to run const checks depending on the arm that one is in. This may seem easy in the given example, but once you involve loops, short circuiting, matching, break and continue, this can get very complex to decide. Additionally even the simple version will complicate the const qualif check a lot, making one of the most fragile things keeping Rust's safety guarantees even more fragile. If there is a less complex version (and I believe there is), we should go for that one.

Reply:
I see. Yes, the "boolean intrinsic" approach implemented here would require the constant evaluator to only try to prove things about the branches that constant evaluation actually can reach. I can see how that could get nasty.

IIUC, what you are proposing is something like: 

```rust
mod intrinsics {
    pub fn amazing_name<..cool type bounds..>(called_in_const: T, called_at_rt: T) -> U
        where ..some cool bounds..;
}
```

that can be used to rewrite the example above as follows:

```rust
const fn eq_ct(x: [i32; 4], y: [i32; 4]) -> bool {
    x[0] == y[0] && x[1] == y[1] && x[2] == y[2] && x[3] == y[3]
}

fn eq_rt(x: [i32; 4], y: [i32; 4]) -> bool {
    let x = _mm_loadu_si128(&x as *const _ as *const _);
    let y = _mm_loadu_si128(&y as *const _ as *const _);
    let r = _mm_cmpeq_epi32(x, y);
    let r = _mm_movemask_ps(transmute(r) );
    r == 0b1111
}

const fn eq(x: [i32; 4], y: [i32; 4]) -> bool {
    amazing_name(|| eq_ct(x, y), || eq_rt(x, y))
}
```

?

Is that correct? Maybe you could post about it in the tracking issue (https://github.com/rust-rfcs/const-eval/issues/7) ? Your last comment there appeared to be in agreement with using a "boolean intrinsic", but it does have quite a bit of issues, so maybe we can move the discussion there.

Reply:
While closures would be neat, they seem an unnecessary complication of such a low level function

```rust
mod intrinsics {
    pub fn const_eval_select<ARG, F, G, RET>(arg: ARG called_in_const: F, called_at_rt: G) -> RET;
}
```

you could then use it as

```rust
const fn eq_ct((x, y): ([i32; 4], [i32; 4])) -> bool {
    x[0] == y[0] && x[1] == y[1] && x[2] == y[2] && x[3] == y[3]
}

fn eq_rt((x, y): ([i32; 4], [i32; 4])) -> bool {
    let x = _mm_loadu_si128(&x as *const _ as *const _);
    let y = _mm_loadu_si128(&y as *const _ as *const _);
    let r = _mm_cmpeq_epi32(x, y);
    let r = _mm_movemask_ps(transmute(r) );
    r == 0b1111
}

const fn eq(x: [i32; 4], y: [i32; 4]) -> bool {
    const_eval_select((x, y), eq_ct, eq_rt)
    // unclear to me how to make these closures work well here
}
```

The `const_eval_select` function then compiles down to a call to `eq_rt(arg)` in llvm and is interpreted as a call to `eq_ct(arg)` in const eval

Reply:
This sounds great.

So how do I implement this? IIUC, I just add an intrinsic to typeck like I did here, and implement it with the "const semantics" in const-eval, and then add a different implementation for it in the llvm backend, and that's it ?

I add an intrinsic like i've done here

Reply:
Yes. Though the llvm backend impl may be a bit involved. Look around for other intrinsics that lower to function calls.

Reply:
Also, since there are no generic bounds, some documentation on the intrinsic on what is valid would be good. Like that the functions given must have the correct argument and return types and that the first function must be a const fn.

Ideally the intrinsic checks would report errors on any illegal uses, but I don't know how easy that is to do

Reply:
> Also, since there are no generic bounds, some documentation on the intrinsic on what is valid would be good.

Sure. The current docs on `is_const_eval` already tackle this, but for this other intrinsic it might suffice that the "run-time function" must have the same operational semantics as the `const fn`.

And maybe elaborate a bit on what that means right now, but as `const fn` evolves, that definition evolves as well. For example, it could explain that by saying that, e.g., for all inputs, the run-time function must return the same values and have the same side-effects as the `const fn`, and that the run-time function can only access the same memory that the `const fn` would be able to access, etc.

Reply:
> Ideally the intrinsic checks would report errors on any illegal uses, but I don't know how easy that is to do

I can't imagine this is feasible, we'd need to analyze the "runtime code" and figure out if it does something non-const friendly, but we can't really tell much about that because the point of this feature is to be able to call SIMD intrinsics or inline assembly, which is essentially stuf we cannot reason about. 

I'd be fine with saying that this intrinsic is `unsafe` to use, and making the user responsible for using it correctly.

Reply:
Oh that's not what I meant. I just mean that

* the signature of the two functions needs to match,
* the first function needs to be `const fn`,
* the return type of both functions is the same as the generic `RET` parameter
* the argument of both functions is the same type as the `ARG` parameter

The content of the functions is indeed left to the user as per the docs you already provided

Reply:
Ah, gotcha. Hm, what about just saying:

```rust
mod intrinsics {
    pub fn const_eval_select<ARG, F, G, RET>(
        arg: ARG called_in_const: F, called_at_rt: G
    ) -> RET 
    where F: FnOnce(ARG) -> RET, G: FnOnce(ARG) -> RET
}
```

?

That way, `typeck` enforces that the arguments and return types match, and well, that they are function types. I don't know how to check that the first function is `const` in  typeck though, but maybe it is possible to do during const qualif.

Reply:
I thought we wouldn't go for closures, to keep things simple? Then it would just be
```rust
pub fn const_eval_select<ARG, RET>(arg: ARG, called_in_const: fn(ARG) -> RET, called_at_rt: fn(ARG) -> RET) -> RET
```

Reply:
function pointers are not the same thing as the zst arguments and won't do us any favours here. The intrinsic would not be able to figure out which function to call and would have to insert function pointer calls and we can't have const function pointers. While all solvable problems, I'd rather not go down that road

I mean you can add `Fn` bounds as you suggestd, that should suffice indeed.

Still needs an extra check to make sure only functions are used as generic arguments